packages/backend/src/open_payments/auth/middleware.ts

```diff
@@ -37,26 +38,45 @@ export function createTokenIntrospectionMiddleware({
         return
       }
       const authService = await ctx.container.use('authService')
-      const grant = await authService.introspect(token)
-      if (!grant) {
+      const tokenInfo = await authService.introspect(token)
+      if (!tokenInfo) {
         ctx.throw(401, 'Invalid Token')
       }
-      const access = grant.findAccess({
-        type,
-        action,
-        identifier: ctx.paymentPointer.url
-      })
+      // TODO
+      // https://github.com/interledger/rafiki/issues/835
+      const access = tokenInfo.access.find(
+        (access) =>
+          access.type === type &&
+          (!access['identifier'] ||
+            access['identifier'] === ctx.paymentPointer.url) &&
+          access.actions.find((tokenAction) => {
+            if (tokenAction == action) {
+              // Unless the relevant grant action is ReadAll/ListAll add the
+              // clientId to ctx for Read/List filtering
```

Comment:
I'm not sure if I understand the comment fully, but it reads as though

```ts
return (
    (action === AccessAction.Read &&
      tokenAction == AccessAction.ReadAll) ||
     (action === AccessAction.List &&
       tokenAction == AccessAction.ListAll)
)
```

should be before 
```
if (tokenAction == action) { ... }
```

Reply:
`action` will never be `ReadAll` or `ListAll`
https://github.com/interledger/rafiki/blob/4db6f8e8ee6e4812af452af379855e4c3ff8d2f6/packages/backend/src/app.ts#L279-L296
so this should be fine, but I'll update the types to make that explicit.

Reply:
:thinking: Actually this is still dependent on `Read` and `ReadAll` (and `List`/`ListAll`) to be mutually exclusive, which is no longer the case (in OpenAPI)
- https://github.com/interledger/open-payments/pull/215